syntax = "proto3";

package masterworker;

service WorkerService {
    rpc DoMap (MapRequest) returns (MapReply) {
    }
    rpc DoReduce (ReduceRequest) returns (ReduceReply) {
    }
}

message ShardSegment {
    string file_name = 1;
    int32 begin = 2;
    int32 end = 3;
}

message ShardInfo {
    int32 shard_id = 1;
    repeated ShardSegment segments = 2;
}

message MapRequest {
    string user_id = 1;
    int32 num_outputs = 2;
    string output_dir = 3;
    ShardInfo shard = 4;
}

message MapReply {
    repeated string file_names = 1;
}

message ReduceRequest {
    int32 reduce_id = 1;
    string user_id = 2;
    int32 num_outputs = 4;
    string output_dir = 5;
    repeated string file_names = 6; // interim file name
}

message ReduceReply {
    repeated string file_names = 1; // final file names
}
